pkg/kubelet/cm/devicemanager/manager_test.go

```diff
@@ -669,6 +670,106 @@ type TestResource struct {
 	topology         bool
 }
 
+func TestFilterByAffinity(t *testing.T) {
+	as := require.New(t)
+	allDevices := ResourceDeviceInstances{
+		"res1": map[string]pluginapi.Device{
+			"dev1": {
+				ID: "dev1",
+				Topology: &pluginapi.TopologyInfo{
+					Nodes: []*pluginapi.NUMANode{
+						{
+							ID: 1,
+						},
+					},
+				},
+			},
+			"dev2": {
+				ID: "dev2",
+				Topology: &pluginapi.TopologyInfo{
+					Nodes: []*pluginapi.NUMANode{
+						{
+							ID: 1,
+						},
+						{
+							ID: 2,
+						},
+					},
+				},
+			},
+			"dev3": {
+				ID: "dev3",
+				Topology: &pluginapi.TopologyInfo{
+					Nodes: []*pluginapi.NUMANode{
+						{
+							ID: 2,
+						},
+					},
+				},
+			},
+			"dev4": {
+				ID: "dev4",
+				Topology: &pluginapi.TopologyInfo{
+					Nodes: []*pluginapi.NUMANode{
+						{
+							ID: 2,
+						},
+					},
+				},
+			},
+		},
+	}
+
+	testManager := ManagerImpl{
+		topologyAffinityStore: newFakeTopologyManager(),
+		allDevices:            allDevices,
+	}
+
+	testCases := []struct {
+		available               sets.String
+		fromAffinityExpected    sets.String
+		notFromAffinityExpected sets.String
+		withoutTopologyExpected sets.String
+	}{
+		{
+			available:               sets.NewString("dev1", "dev2"),
+			fromAffinityExpected:    sets.NewString("dev2"),
+			notFromAffinityExpected: sets.NewString("dev1"),
+			withoutTopologyExpected: sets.NewString(),
+		},
+		{
+			available:               sets.NewString("dev1", "dev2", "dev3", "dev4"),
+			fromAffinityExpected:    sets.NewString("dev2", "dev3", "dev4"),
+			notFromAffinityExpected: sets.NewString("dev1"),
+			withoutTopologyExpected: sets.NewString(),
+		},
+	}
+
+	for _, testCase := range testCases {
+		fromAffinity, notFromAffinity, withoutTopology := testManager.filterByAffinity("", "", "res1", testCase.available)
+		as.Truef(fromAffinity.Equal(testCase.fromAffinityExpected), "expect devices from affinity to be %v but got %v", testCase.fromAffinityExpected, fromAffinity)
+		as.Truef(notFromAffinity.Equal(testCase.notFromAffinityExpected), "expect devices not from affinity to be %v but got %v", testCase.notFromAffinityExpected, notFromAffinity)
+		as.Truef(withoutTopology.Equal(testCase.withoutTopologyExpected), "expect devices without topology to be %v but got %v", testCase.notFromAffinityExpected, notFromAffinity)
+	}
+}
+
+type fakeTopologyManager struct {
+	topologymanager.Manager
+}
+
+func newFakeTopologyManager() topologymanager.Manager {
+	return &fakeTopologyManager{}
+}
+
+func (m *fakeTopologyManager) GetAffinity(podUID string, containerName string) topologymanager.TopologyHint {
+	// return hint with numa2 set
+	affinity, _ := bitmask.NewBitMask(2)
+	return topologymanager.TopologyHint{
+		NUMANodeAffinity: affinity,
+		Preferred:        true,
+	}
+}
+
```

Comment:
I think this is a bit to specific to just this use case with the hard-coding of a single hint with an affinity of `2` deep in the `GetAffinity()` call. If you want to build a `fakeTopologyManager` such as this, we should try and make it a bit more general so that it could be used by other test cases as well.

Reply:
@klueska just update the unit test code, PTAL

Reply:
Another way might be reuse the existing topologymanager.NewFakeManager by passing a fake Hint as its parameter and store in topologymanager.fakeManager{}, but that would require more existing code to be changed, do you think this is a good idea?

Reply:
@klueska I've update to use existing code in:

https://github.com/kubernetes/kubernetes/blob/1295b2c4b59b8bba58186294eb2d76a7ecb4987e/pkg/kubelet/cm/topologymanager/fake_topology_manager.go#L25-L36